myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/JHSTaskFactoryImpl.java

```diff
@@ -0,0 +1,366 @@
+          String executorPathString = myriadExecutorConfiguration.getPath();
+          String sudoExecutorCmd = "sudo -E -u " + cfg.getFrameworkUser().get() + " -H " +
+              executorCmd;
+
+          //Concatenate all the subcommands
+          String cmd = tarCmd + "&&" + chownCmd + "&&" + configCopyCmd + "&&" + sudoExecutorCmd;
+
+          //get the nodemanagerURI
+          //We're going to extract ourselves, so setExtract is false
+          LOGGER.info("Getting Hadoop distribution from:" + nmURIString);
+          URI nmUri = URI.newBuilder().setValue(nmURIString).setExtract(false)
+              .build();
+
+          //get configs directly from resource manager
+          String configUrlString = getConfigurationUrl();
+          LOGGER.info("Getting config from:" + configUrlString);
+          URI configUri = URI.newBuilder().setValue(configUrlString)
+              .build();
+
+          //get the executor URI
+          LOGGER.info("Getting executor from:" + executorPathString);
+          URI executorUri = URI.newBuilder().setValue(executorPathString).setExecutable(true)
+              .build();
+
+          LOGGER.info("Slave will execute command:" + cmd);
+          commandInfo.addUris(nmUri).addUris(configUri).addUris(executorUri).setValue("echo \"" + cmd + "\";" + cmd);
+
+          commandInfo.setUser(cfg.getFrameworkSuperUser().get());
+
+        } else {
+          commandInfo.setValue(executorCmd);
+        }
+        return commandInfo.build();
+      }
+      
+      private static String getFileName(String uri) {
+        int lastSlash = uri.lastIndexOf('/');
+        if (lastSlash == -1) {
+          return uri;
+        } else {
+          String fileName = uri.substring(lastSlash + 1);
+          Preconditions.checkArgument(!Strings.isNullOrEmpty(fileName),
+              "URI should not have a slash at the end");
+          return fileName;
+        }
+      }
+
+      private String getConfigurationUrl() {
+        YarnConfiguration conf = new YarnConfiguration();
+        String httpPolicy = conf.get(YARN_HTTP_POLICY);
+        if (httpPolicy != null && httpPolicy.equals(YARN_HTTP_POLICY_HTTPS_ONLY)) {
+          String address = conf.get(YARN_RESOURCEMANAGER_WEBAPP_HTTPS_ADDRESS);
+          if (address == null || address.isEmpty()) {
+            address = conf.get(YARN_RESOURCEMANAGER_HOSTNAME) + ":8090";
+          }
+          return "https://" + address + "/conf";
+        } else {
+          String address = conf.get(YARN_RESOURCEMANAGER_WEBAPP_ADDRESS);
+          if (address == null || address.isEmpty()) {
+            address = conf.get(YARN_RESOURCEMANAGER_HOSTNAME) + ":8088";
+          }
+          return "http://" + address + "/conf";
+        }
+      }
+
+      /**
+       * Helper method to reserve ports
+       * @param offer
+       * @param requestedPorts
+       * @return
+       */
+      private List<Long> getAvailablePorts(Offer offer, List<Long> requestedPorts) {
```

Comment:
Changed behavior with ports - see comments with the latest commit:

 New behavior: if ports are specified in the configuration they will be used irrespective of anything. If ports are not specified in configuration they will be selected from list of ports from Mesos offer.

In both cases following properties will be set:
        "myriad.mapreduce.jobhistory.admin.address",
        "myriad.mapreduce.jobhistory.address",
        "myriad.mapreduce.jobhistory.webapp.address"

So they can be used in configuration files. E.g:
<property>
  <name>mapreduce.jobhistory.admin.address</name>
  <value>${myriad.mapreduce.jobhistory.admin.address}</value>
</property>
<property>
  <name>mapreduce.jobhistory.address</name>
  <value>${myriad.mapreduce.jobhistory.address}</value>
</property>
<property>
  <name>mapreduce.jobhistory.webapp.address</name>
  <value>${myriad.mapreduce.jobhistory.webapp.address}</value>
</property>

In case of random ports JHS web app address  has to be known by clients in order to successfully start/finish job.